apps/cowswap-frontend/src/modules/trade/hooks/setupTradeState/useSetupTradeState.ts

```diff
@@ -44,9 +61,71 @@ export function useSetupTradeState(): void {
 
   const isAlternativeModalVisible = useIsAlternativeOrderModalVisible()
 
+  // Force update the URL to match provider chainId after network change
+  const forceUrlUpdateAfterNetworkChange = useCallback(
+    (targetChainId: SupportedChainId) => {
+      // Clear any existing timeout
+      if (forceUrlUpdateTimeoutRef.current) {
+        clearTimeout(forceUrlUpdateTimeoutRef.current)
+        forceUrlUpdateTimeoutRef.current = null
+      }
+
+      // Set a timeout to ensure all state updates have propagated before forcing URL update
+      forceUrlUpdateTimeoutRef.current = setTimeout(() => {
+        // Only update if we're still on the same provider chain
+        if (targetChainId === providerChainId) {
+          urlUpdateRetryCount.current = 0
+
+          // Use default tokens when switching chains
+          const defaultState = getDefaultTradeRawState(targetChainId)
+          tradeNavigate(targetChainId, defaultState)
+
+          // For TWAP mode, immediately update the state as well to ensure consistency
+          if (updateState) {
+            updateState({ ...defaultState })
+          }
+        }
+
+        // Clear pending network change state
+        setPendingNetworkChange(null)
+        forceUrlUpdateTimeoutRef.current = null
+      }, NETWORK_CHANGE_TIMEOUT)
+    },
+    [providerChainId, tradeNavigate, updateState],
+  )
+
+  // Monitor URL state for reversions and force correction if needed
+  useEffect(() => {
+    // Only for TWAP mode and when we've had a successful provider change
+    if (isTwapMode && providerChainId && urlChainId && providerChainId !== urlChainId) {
+      // If URL has reverted to old chainId after we thought we updated it
+      if (pendingNetworkChange === null && forceUrlUpdateTimeoutRef.current === null) {
+        // Only retry a limited number of times to avoid infinite loops
+        if (urlUpdateRetryCount.current < MAX_URL_UPDATE_RETRIES) {
+          urlUpdateRetryCount.current++
+          forceUrlUpdateAfterNetworkChange(providerChainId)
+        } else {
+          console.error('[TRADE STATE]', 'Exceeded max retries for URL correction', {
+            providerChainId,
+            urlChainId,
+          })
+        }
+      }
+    } else {
+      // Reset retry counter when chain IDs match
+      urlUpdateRetryCount.current = 0
+    }
+  }, [providerChainId, urlChainId, isTwapMode, pendingNetworkChange, forceUrlUpdateAfterNetworkChange])
+
```

Comment:
I don't think this is the right way to go.

For one, it does not happen with other trade modes, does it? What's special about TWAP that's breaking?
Second, having this timeout + retry logic seems overkill.
Third, the current support for TWAPs is only for Safes which (most of the time) are only ever present on one chain.
To change the network we'd need to change the Safe, which I think would cause a whole app refresh, no?

If we do want to apply the same behaviour, I think we just need to load the hook that does that for swap/limit.

Reply:
@alfetopito I've refactored code now, please review